File: native/ios/FireballNative/Core/ItunesFeedParser.swift
import Foundation

enum ItunesFeedParser {
    struct ChartEntry {
        let id: String
        let title: String
        let artist: String
        let artwork: String?
        let previewUrl: String?
    }

    static func feedEntries(from data: Data) -> [[String: Any]] {
        guard let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let feed = root["feed"] as? [String: Any],
              let entry = feed["entry"] else {
            return []
        }
        if let list = entry as? [[String: Any]] { return list }
        if let one = entry as? [String: Any] { return [one] }
        return []
    }

    static func topCharts(from entries: [[String: Any]]) -> [ChartEntry] {
        entries.compactMap { e in
            guard let idObj = e["id"] as? [String: Any],
                  let attrs = idObj["attributes"] as? [String: Any],
                  let id = attrs["im:id"] as? String,
                  !id.isEmpty else {
                return nil
            }
            let title = (e["im:name"] as? [String: Any])?["label"] as? String ?? "—"
            let artist = (e["im:artist"] as? [String: Any])?["label"] as? String ?? "—"
            let images = e["im:image"] as? [[String: Any]]
            let artwork = images?.dropFirst(2).first?["label"] as? String ?? images?.last?["label"] as? String
            return ChartEntry(
                id: id,
                title: title,
                artist: artist,
                artwork: artwork,
                previewUrl: extractPreviewUrl(e["link"])
            )
        }
    }

    private static func extractPreviewUrl(_ link: Any?) -> String? {
        if let list = link as? [[String: Any]] {
            for item in list {
                guard let attrs = item["attributes"] as? [String: Any] else { continue }
                let type = attrs["type"] as? String
                let title = attrs["title"] as? String
                if type == "audio/x-m4a" || title == "Preview" {
                    return attrs["href"] as? String
                }
            }
            return nil
        }
        if let map = link as? [String: Any],
           let attrs = map["attributes"] as? [String: Any] {
            return attrs["href"] as? String
        }
        return nil
    }
}

File: native/android/core-data/src/main/kotlin/com/fireball/nativeapp/core/data/ItunesFeedParser.kt
package com.fireball.nativeapp.core.data

import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

/** Parses Apple Music RSS JSON feeds (same shape as Flutter [FireballApi.appleRssFeedEntries]). */
object ItunesFeedParser {
    fun feedEntries(root: JsonObject): List<JsonObject> {
        val entry = root["feed"]?.jsonObject?.get("entry") ?: return emptyList()
        return when (entry) {
            is JsonArray -> entry.mapNotNull { it.jsonObject }
            is JsonObject -> listOf(entry)
            else -> emptyList()
        }
    }

    data class ChartEntry(
        val id: String,
        val title: String,
        val artist: String,
        val artwork: String?,
        val previewUrl: String?,
    )

    fun topChartsFromFeed(entries: List<JsonObject>): List<ChartEntry> =
        entries.mapNotNull { e ->
            val id = e["id"]?.jsonObject
                ?.get("attributes")?.jsonObject
                ?.get("im:id")?.jsonPrimitive?.content
                ?: return@mapNotNull null
            if (id.isBlank()) return@mapNotNull null
            ChartEntry(
                id = id,
                title = e["im:name"]?.jsonObject?.get("label")?.jsonPrimitive?.content ?: "—",
                artist = e["im:artist"]?.jsonObject?.get("label")?.jsonPrimitive?.content ?: "—",
                artwork = e["im:image"]?.jsonArray?.getOrNull(2)?.jsonObject?.get("label")?.jsonPrimitive?.content,
                previewUrl = extractPreviewUrl(e["link"]),
            )
        }

    private fun extractPreviewUrl(link: JsonElement?): String? {
        if (link == null) return null
        if (link is JsonArray) {
            for (item in link) {
                val obj = item.jsonObject
                val attrs = obj["attributes"]?.jsonObject ?: continue
                val type = attrs["type"]?.jsonPrimitive?.content
                val title = attrs["title"]?.jsonPrimitive?.content
                if (type == "audio/x-m4a" || title == "Preview") {
                    return attrs["href"]?.jsonPrimitive?.content
                }
            }
            return null
        }
        if (link is JsonObject) {
            return link["attributes"]?.jsonObject?.get("href")?.jsonPrimitive?.content
        }
        return null
    }
}
